src/server/auth/server/oidc.go

```diff
@@ -192,23 +256,25 @@ func (a *apiServer) handleExchange(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	si := stateInfoMap[state]
-	if idToken.Nonce != si.Nonce {
-		logrus.Errorf("expected nonces to match, instead set nonce %v but got nonce %v", si.Nonce, idToken.Nonce)
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
-		return
-	}
-	logrus.Infof("nonce is %v", idToken.Nonce)
-
-	si.Token = tok.AccessToken
-	logrus.Infof("saving state with access token")
-	stateInfoMap[state] = si
+	// Verify the ID token, and if it's valid, add it to this state's SessionInfo
+	// in etcd, so that any concurrent Authorize() calls can discover it and give
+	// the caller a Pachyderm token.
+	_, err = col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
+		var si auth.SessionInfo
+		return sp.States.ReadWrite(stm).Update(state, &si, func() error {
+			if idToken.Nonce != si.Nonce {
```

Comment:
This is the only place we used the ID token we get from the IDP. Can discard after this.